Add global cycle pos cmds and show it in gui

sebageek opened the pull request:
This commit adds the commands "global_cycle_len" and "global_cycle_pos"
to sooperlooper and a gui element showing the position in the current
sooperlooper cycle. This feature only works in Internal or Jack sync
mode with quantize set to cycle or 8th (i.e. whenever sooperlooper
actually has an internal cycle).

When recording the first loop in sync internal + quantize cycle + first
loop on sync it is hard to hit the first note at the beginning of the
cycle. A visual indicator of where we are in the cycle helps to be on
time in situation where no other cue is available.

The gui element has currently no text and only displays the current
cycle if its length is >= 1s. Both of these measures have been taken to
reduce visual noise in the gui.

---

Soo, I've been having this problem with not knowing where in the sooperlooper cycle I was when I wanted to record my first loop, so I decided to make it visible. Display is done with a SliderBar, which I think is quite neat for displaying this (..or it's the best widget I've found on short notice to do the job). I hope this is useful to someone else as well.

This is my first deep dive into the sooperlooper source and I tried to orient myself a bit on the code surrounding my changes. Also, if you think the new widget in the top bar creates too much visual noise here's a couple things we could still implement:
 * hide the widget when we're not in sync mode jack/internal and when the global cycle is shorter than 1s
 * make display of this widget configurable via preferences

Though I had some trouble with hiding/displaying that widget. Reviews / comments welcome! ;)

ericfont commented:
I'm trying out.  My suggestion would be that instead of displaying a new widget, what about if the button for "quantize" instead becomes this display when the button is set to "cycle".  Just my two cents.

sebageek commented:
I'll take a look at putting this into the global values. My first attempt of this did not work, but it might be because I haven't properly requested updates in the GUI.

Regarding the GUI element.... I like the idea as well, though the quantize box is a `ChoiceBox` and doesn't have the relevant indicator code that `SliderBar` has. I don't think I can make it a `SliderBar`. I could try to port the indicator code to `ChoiceBox` if that would be an OK use-case.

sebageek commented:
I taught the `ChoiceBox` how to draw a bar! The interface is far less sophisticated in comparison to the `SliderBar` (it only displays values between 0.0 and 1.0, no indicator and all of that), but it does the trick (I think in this case it even looks better without the indicator).

![image](https://user-images.githubusercontent.com/2045907/138006231-7f5df260-8c65-4965-bdd0-f63b824d4e19.png)

@essej could you take a look at the code in the second commit (b622663) and see if the changes in it are acceptable to you? If they are I'd go ahead, squash the two commits and remove the extra `SliderBar` I added for the cycle in the first commit.

ericfont commented:
I'm confused because when I first start your branch fresh, it looks like:

![image](https://user-images.githubusercontent.com/6502474/138134973-df7243ad-a8e8-43a9-b2c9-1e11242b9c2c.png)

I don't think I'm supposed to see that `curr cycle` box in the top right.

ericfont commented:
oh but it seems to work with those settings your provided me.  Though I think now the `curr. cycle` box in the ttop right is no longer necessary since you got the `quantize` box working with animation.

ericfont commented:
I notice that on 120bpm, if I set 8th/cycle to 4.0 the display work, but if I reduce 8th/cycle to 3.0 then there is no animation.  So maybe the reason I didn't animation before was because I just had the setting below 4.0 earlier.

sebageek commented:
As written in https://github.com/essej/sooperlooper/pull/26#issuecomment-947195358 I first wanted to ask if that API change looks right before I remove the other commit, but on the other hand... I can just prepare the commit in the way it's meant to be merged later.

sebageek commented:
> I notice that on 120bpm, if I set 8th/cycle to 4.0 the display work, but if I reduce 8th/cycle to 3.0 then there is no animation.
Yes, there is a limit in the code [here](https://github.com/essej/sooperlooper/pull/26/files#diff-28dbbb4cd1ccf57ea15d5d6972986b442bf095fa6e178f6a0e520fc2d02e69e1R758). I only display the status if the global cycle is >= 1s, as for less than a second it just creates visual noise and is not really helpful. If you want try it out! Just comment out that one `if` and it should give you a very fast animation. :)

ericfont commented:
I'm noticing funny behavior which might be a separate issues, but when I have tempo at 59 bpm or lower, then pressing the up/down control on the tempo box causes the number in `8th/cycle` to double, thus easily reaching astronomical values.

sebageek commented:
Okay, squashed the commits and rewrote the commit message a bit so it reflects the current state + the ChoiceBox API change.
I also tested this again with sync to Jack, where this feature doesn't work. That is most likely because sooperlooper just listens to the sync/start event on Jack. We could also display this cycle by measuring time between two sync events, but I think that is something for a future PR. I also write into the commit message that it only works with internal sync for now.

jean-emmanuel commented:
@sebageek is global_cycle_pos controllable with osc messages ? It feels it should be, but I can't manage to do it.
